### PR TITLE
perf: loop optimization

### PR DESCRIPTION
**Pull request by Vetlix**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/contentTypeParser.js [hunk 1]
```diff
@@ -93,19 +93,19 @@ ContentTypeParser.prototype.existingParser = function (contentType) {
 }
 
 ContentTypeParser.prototype.getParser = function (contentType) {
-  /* eslint-disable no-var */
-  for (var i = 0; i < this.parserList.length; i++) {
-    if (contentType.indexOf(this.parserList[i]) > -1) {
-      const parser = this.customParsers[this.parserList[i]]
+  for (let i = 0, len = this.parserList.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/contentTypeParser.js:96]: please keep it a var

File: lib/contentTypeParser.js [hunk 2]
```diff
@@ -93,19 +93,19 @@ ContentTypeParser.prototype.existingParser = function (contentType) {
...
       this.cache.set(contentType, parser)
       return parser
     }
   }
 
-  /* eslint-disable no-var */
-  for (var j = 0; j < this.parserRegExpList.length; j++) {
-    if (this.parserRegExpList[j].test(contentType)) {
-      const parser = this.customParsers[this.parserRegExpList[j]]
+  for (let i = 0, len = this.parserRegExpList.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/contentTypeParser.js:105]: please keep it a var

File: lib/context.js [hunk 3]
```diff
@@ -31,8 +31,8 @@ function Context (schema, handler, Reply, Request, contentTypeParser, config, er
 function defaultSchemaErrorFormatter (errors, dataVar) {
   let text = ''
   const separator = ', '
-  /* eslint-disable no-var */
-  for (var i = 0; i < errors.length; i++) {
+
+  for (let i = 0, len = errors.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/context.js:35]: keep it a var

File: lib/decorate.js [hunk 4]
```diff
@@ -64,7 +64,7 @@ function checkReplyExistence (name) {
 }
 
 function checkDependencies (instance, deps) {
-  for (let i = 0; i < deps.length; i++) {
+  for (let i = 0, len = deps.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/decorate.js:67]: same

File: lib/fourOhFour.js [hunk 5]
```diff
@@ -132,11 +132,12 @@ function fourOhFour (options) {
...
-      for (const hook of lifecycleHooks) {
-        const toSet = this[kHooks][hook]
-          .concat(opts[hook] || [])
+      for (let i = 0, len = lifecycleHooks.length; i !== len; ++i) {
+        const lifecycleHook = lifecycleHooks[i]
+        const toSet = this[kHooks][lifecycleHook]
+          .concat(opts[lifecycleHook] || [])
           .map(h => h.bind(this))
-        context[hook] = toSet.length ? toSet : null
+        context[lifecycleHook] = toSet.length ? toSet : null
```
Comment (mcollina) [inline lib/fourOhFour.js:140]: this whole block is not a hotpath, do not optimize

File: lib/initialConfigValidation.js [hunk 6]
```diff
@@ -19,7 +19,8 @@ function validateInitialConfig (options) {
 function deepFreezeObject (object) {
   const properties = Object.getOwnPropertyNames(object)
 
-  for (const name of properties) {
+  for (let i = 0, len = properties.length; i !== len; ++i) {
+    const name = properties[i]
```
Comment (mcollina) [inline lib/initialConfigValidation.js:23]: not a hotpath, don't optimize

File: lib/logger.js [hunk 7]
```diff
@@ -107,8 +107,10 @@ function isValidLogger (logger) {
 
   let result = true
   const methods = ['info', 'error', 'debug', 'fatal', 'warn', 'trace', 'child']
-  for (let i = 0; i < methods.length; i += 1) {
-    if (!logger[methods[i]] || typeof logger[methods[i]] !== 'function') {
+
+  for (let i = 0; i !== methods.length; ++i) {
```
Comment (mcollina) [inline lib/logger.js:111]: not a hotpath, do not optimize

File: lib/pluginOverride.js [hunk 8]
```diff
@@ -57,7 +57,8 @@ module.exports = function override (old, fn, opts) {
     instance[kFourOhFour].arrange404(instance)
   }
 
-  for (const hook of instance[kHooks].onRegister) hook.call(this, instance, opts)
+  const hooks = instance[kHooks].onRegister
+  for (let i = 0, len = hooks.length; i !== len; hooks[i++].call(this, instance, opts));
```
Comment (mcollina) [inline lib/pluginOverride.js:61]: do not optimize

File: lib/pluginUtils.js [hunk 9]
```diff
@@ -20,9 +20,10 @@ function getPluginName (func) {
   const cache = require.cache
   const keys = Object.keys(cache)
 
-  for (let i = 0; i < keys.length; i++) {
-    if (cache[keys[i]].exports === func) {
-      return keys[i]
+  for (let i = 0, len = keys.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/pluginUtils.js:23]: I don't think this is a hotpath

File: lib/reply.js [hunk 10]
```diff
@@ -244,8 +244,9 @@ Reply.prototype.header = function (key, value) {
 Reply.prototype.headers = function (headers) {
   const keys = Object.keys(headers)
   /* eslint-disable no-var */
-  for (var i = 0; i < keys.length; i++) {
-    this.header(keys[i], headers[keys[i]])
+  for (let i = 0, len = keys.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/reply.js:247]: keep it a var

File: lib/route.js [hunk 11]
```diff
@@ -130,10 +130,10 @@ function buildRouting (options) {
     throwIfAlreadyStarted('Cannot add route when fastify instance is already started!')
 
     if (Array.isArray(opts.method)) {
-      /* eslint-disable no-var */
-      for (var i = 0; i < opts.method.length; i++) {
-        if (supportedMethods.indexOf(opts.method[i]) === -1) {
-          throw new Error(`${opts.method[i]} method is not supported!`)
+      for (let i = 0, len = opts.method.length; i < len; ++i) {
```
Comment (mcollina) [inline lib/route.js:133]: keep it a var

File: lib/route.js [hunk 12]
```diff
@@ -156,7 +156,7 @@ function buildRouting (options) {
 
     this.after((notHandledErr, done) => {
       const path = opts.url || opts.path
-      if (path === '/' && prefix.length > 0 && opts.method !== 'HEAD') {
+      if (path === '/' && prefix.length && opts.method !== 'HEAD') {
```
Comment (mcollina) [inline lib/route.js:159]: is this faster?

File: lib/route.js [hunk 13]
```diff
@@ -209,9 +209,10 @@ function buildRouting (options) {
 
       if (prefixing === false) {
       // run 'onRoute' hooks
-        for (const hook of this[kHooks].onRoute) {
+        const hooks = this[kHooks].onRoute
+        for (let i = 0, len = hooks.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/route.js:213]: no need to optimize

File: lib/route.js [hunk 14]
```diff
@@ -270,7 +271,8 @@ function buildRouting (options) {
       // the route registration. To be sure to also load those hooks,
       // we must listen for the avvio's preReady event, and update the context object accordingly.
       avvio.once('preReady', () => {
-        for (const hook of lifecycleHooks) {
+        for (let i = 0, len = lifecycleHooks.length; i !== len; ++i) {
```
Comment (mcollina) [inline lib/route.js:274]: no need to optimize

File: lib/schema-compilers.js [hunk 15]
```diff
@@ -37,16 +37,15 @@ function ValidatorCompiler (externalSchemas, options, cache) {
     nullable: true
   }, options.customOptions, { cache }))
 
-  if (options.plugins && options.plugins.length > 0) {
-    for (const plugin of options.plugins) {
+  if (options.plugins && options.plugins.length) {
+    for (let i = 0, len = options.plugins.length; i !== len; ++i) {
+      const plugin = options.plugins[i]
```
Comment (mcollina) [inline lib/schema-compilers.js:42]: no need to optimize

File: lib/schemas.js [hunk 16]
```diff
@@ -60,7 +60,9 @@ function normalizeSchema (routeSchemas) {
   generateFluentSchema(routeSchemas)
 
   // let's check if our schemas have a custom prototype
-  for (const key of ['headers', 'querystring', 'params', 'body']) {
+  const keys = ['headers', 'querystring', 'params', 'body']
+  for (let i = 0, len = keys.length; i !== len; ++i) {
+    const key = keys[i]
```
Comment (mcollina) [inline lib/schemas.js:65]: no need to optimize

File: lib/schemas.js [hunk 17]
```diff
@@ -94,17 +97,27 @@ function normalizeSchema (routeSchemas) {
...
 
 function generateFluentSchema (schema) {
-  for (const key of SCHEMAS_SOURCE) {
-    if (schema[key] && (schema[key].isFluentSchema || schema[key][kFluentSchema])) {
-      schema[key] = schema[key].valueOf()
+  // for (const key of SCHEMAS_SOURCE) {
+  //   if (schema[key] && (schema[key].isFluentSchema || schema[key][kFluentSchema])) {
+  //     schema[key] = schema[key].valueOf()
+  //   }
+  // }
```
Comment (mcollina) [inline lib/schemas.js:104]: why is this commented out?

File: lib/schemas.js [hunk 18]
```diff
@@ -94,17 +97,27 @@ function normalizeSchema (routeSchemas) {
...
+
+  for (let i = 0, len = SCHEMAS_SOURCE.length; i !== len; ++i) {
+    const key = SCHEMAS_SOURCE[i]
+    const keyValue = schema[key]
+    if (keyValue && (keyValue.isFluentSchema || keyValue[kFluentSchema])) {
+      schema[key] = keyValue.valueOf()
     }
```
Comment (mcollina) [inline lib/schemas.js:100]: no need to optimize

File: lib/schemas.js [hunk 19]
```diff
@@ -94,17 +97,27 @@ function normalizeSchema (routeSchemas) {
...
     const httpCodes = Object.keys(schema.response)
-    for (const code of httpCodes) {
-      if (schema.response[code].isFluentSchema || schema.response[code][kFluentSchema]) {
-        schema.response[code] = schema.response[code].valueOf()
+    for (let i = 0, len = httpCodes.length; i !== len; ++i) {
+      const code = httpCodes[i]
+      const codeValue = schema.response[code]
+      if (codeValue.isFluentSchema || codeValue[kFluentSchema]) {
+        schema.response[code] = codeValue.valueOf()
       }
```
Comment (mcollina) [inline lib/schemas.js:108]: no need to optimize

**Review by mcollina — CHANGES_REQUESTED** [review]
Thanks! Have you got some numbers/benchmarks before/after?

File: lib/schemas.js (hunk 17 above)
Comment (Vetlix) [reply lib/schemas.js:104]: Probably leftovers from me editing that I forgot to remove

File: lib/route.js (hunk 12 above)
Comment (Vetlix) [reply lib/route.js:159]: Yes, it should be

**Comment by Vetlix** [comment]
> 
> 
> Thanks! Have you got some numbers/benchmarks before/after?

I do not, it would be nice if someone else could run them for me 🤷

**Comment by RafaelGSS** [comment]
> > Thanks! Have you got some numbers/benchmarks before/after?
> 
> I do not, it would be nice if someone else could run them for me

I do that for you.

**Comment by RafaelGSS** [comment]
```
rafaelgss@rafaelgss-Inspiron-7572:~/repos/os/fastify$ npm run benchmark

> fastify@3.14.2 benchmark /home/rafaelgss/repos/os/fastify
> npx concurrently -k -s first "node ./examples/simple.js" "npx autocannon -c 100 -d 5 -p 10 localhost:3000/"

[1] Running 5s test @ http://localhost:3000/
[1] 100 connections with 10 pipelining factor
[1]
[1] ┌─────────┬───────┬───────┬───────┬────────┬──────────┬─────────┬────────┐
[1] │ Stat    │ 2.5%  │ 50%   │ 97.5% │ 99%    │ Avg      │ Stdev   │ Max    │
[1] ├─────────┼───────┼───────┼───────┼────────┼──────────┼─────────┼────────┤
[1] │ Latency │ 16 ms │ 35 ms │ 66 ms │ 137 ms │ 33.73 ms │ 22.7 ms │ 335 ms │
[1] └─────────┴───────┴───────┴───────┴────────┴──────────┴─────────┴────────┘
[1] ┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
[1] │ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
[1] ├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
[1] │ Req/Sec   │ 15415   │ 15415   │ 32607   │ 33567   │ 29248.8 │ 6946.89 │ 15411   │
[1] ├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
[1] │ Bytes/Sec │ 2.88 MB │ 2.88 MB │ 6.09 MB │ 6.28 MB │ 5.47 MB │ 1.3 MB  │ 2.88 MB │
[1] └───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘
[1]
[1] Req/Bytes counts sampled once per second.
[1]
[1] 146k requests in 5.06s, 27.3 MB read
[1] npx autocannon -c 100 -d 5 -p 10 localhost:3000/ exited with code 0
--> Sending SIGTERM to other processes..
[0] node ./examples/simple.js exited with code SIGTERM
```
This branch
```
rafaelgss@rafaelgss-Inspiron-7572:~/repos/os/fastify$ gh pr checkout 2992
Switched to branch 'loop-optimization'
rafaelgss@rafaelgss-Inspiron-7572:~/repos/os/fastify$ npm run benchmark

> fastify@3.14.2 benchmark /home/rafaelgss/repos/os/fastify
> npx concurrently -k -s first "node ./examples/simple.js" "npx autocannon -c 100 -d 5 -p 10 localhost:3000/"

[1] Running 5s test @ http://localhost:3000/
[1] 100 connections with 10 pipelining factor
[1]
[1] ┌─────────┬───────┬───────┬───────┬────────┬──────────┬──────────┬────────┐
[1] │ Stat    │ 2.5%  │ 50%   │ 97.5% │ 99%    │ Avg      │ Stdev    │ Max    │
[1] ├─────────┼───────┼───────┼───────┼────────┼──────────┼──────────┼────────┤
[1] │ Latency │ 16 ms │ 34 ms │ 72 ms │ 149 ms │ 33.32 ms │ 21.64 ms │ 271 ms │
[1] └─────────┴───────┴───────┴───────┴────────┴──────────┴──────────┴────────┘
[1] ┌───────────┬────────┬────────┬─────────┬────────┬─────────┬─────────┬────────┐
[1] │ Stat      │ 1%     │ 2.5%   │ 50%     │ 97.5%  │ Avg     │ Stdev   │ Min    │
[1] ├───────────┼────────┼────────┼─────────┼────────┼─────────┼─────────┼────────┤
[1] │ Req/Sec   │ 15487  │ 15487  │ 33215   │ 34239  │ 29604   │ 7123.93 │ 15485  │
[1] ├───────────┼────────┼────────┼─────────┼────────┼─────────┼─────────┼────────┤
[1] │ Bytes/Sec │ 2.9 MB │ 2.9 MB │ 6.21 MB │ 6.4 MB │ 5.54 MB │ 1.33 MB │ 2.9 MB │
[1] └───────────┴────────┴────────┴─────────┴────────┴─────────┴─────────┴────────┘
[1]
[1] Req/Bytes counts sampled once per second.
[1]
[1] 148k requests in 5.07s, 27.7 MB read
```

Machine info:
```
OS
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=20
DISTRIB_CODENAME=ulyana
DISTRIB_DESCRIPTION="Linux Mint 20 Ulyana"
NAME="Linux Mint"
VERSION="20 (Ulyana)"

Hardware
Model name:                  Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
CPU(s):                      8
RAM:                         16GB
```

Note: Non-dedicated machine.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Good work!

**Review by RafaelGSS — COMMENTED** [review]
Why this should be more optimized? I mean, in the bytecode likely it will perform equivalent.

**Comment by Vetlix** [comment]
> Why this should be more optimized? I mean, in the bytecode likely it will perform equivalent.

You can view one of the reasons in this issue: https://github.com/nodejs/undici/pull/656#pullrequestreview-632593879

**Comment by tbnritzdoge** [comment]
> Why this should be more optimized? I mean, in the bytecode likely it will perform equivalent.

It's actually not equivalent tho, ++i is slightly faster than i++ and uses less memory as it doesn't make a copy.

**Comment by wzhouwzhou** [comment]
Downvoting 👎 with a degree of agreement to @RafaelGSS: even though I am a strong proponent for hyperefficient enterprise code, in web-based applications and frameworks, the bottleneck for optimisation is very likely to be elsewhere. Code like this fuels a premature optimisation rage that makes public libraries with many moving pieces harder to maintain and code less readable if style is inconsistent and non-easily enforced __consistently__. 

Misleading benchmarks with insufficient explanation such as those above only show v8 turbofan optimisations that kick in at large scales. As we see, even in isolated, seemingly picture-perfect occasions, the almost-the-same averages of 29248 and 29604 reqs/sec with a HUGE standard deviation of 7000 do not prove any statistically significant difference (in fact the now-larger min of 15485 is a nod towards an argument that this PR introduces additional overhead for v8 to manage). **In fact, we have no evidence to reject the null hypothesis here; the test results are inconclusive.**

Rather than PR farming, I strongly recommend we spend some time optimising other known bottlenecked areas of the library first.

**Comment by Vetlix** [comment]
I appreciate the feedback as I'm sort of new to the open-source community. I will take what you said into consideration.

Meanwhile, I would like to make everyone elses day by linking this comment section: https://github.com/nodejs/undici/pull/656#issuecomment-817174845

File: lib/route.js (hunk 12 above)
Comment (zekth) [reply lib/route.js:160]: I just run few benches here are the results: (case 1 is main case 2 is this branch)
```
case 1 x 793,311,168 ops/sec ±2.07% (90 runs sampled)
case 2 x 838,535,224 ops/sec ±0.84% (94 runs sampled)
case 1 x 813,989,092 ops/sec ±2.55% (92 runs sampled)
case 2 x 799,507,815 ops/sec ±1.45% (90 runs sampled)
case 1 x 812,507,713 ops/sec ±2.10% (86 runs sampled)
case 2 x 848,921,418 ops/sec ±1.11% (96 runs sampled)
case 1 x 854,137,890 ops/sec ±0.48% (96 runs sampled)
case 2 x 857,603,147 ops/sec ±0.17% (95 runs sampled)
case 1 x 859,320,269 ops/sec ±0.21% (94 runs sampled)
case 2 x 857,643,300 ops/sec ±0.65% (96 runs sampled)
```

File: lib/pluginUtils.js [hunk 20]
```diff
@@ -21,8 +21,9 @@ function getPluginName (func) {
   const keys = Object.keys(cache)
 
   for (let i = 0; i < keys.length; i++) {
-    if (cache[keys[i]].exports === func) {
-      return keys[i]
+    const key = keys[i]
```
Comment (zekth) [inline lib/pluginUtils.js:24]: what is the optimization in here?

Comment (Vetlix) [reply lib/pluginUtils.js:24]: Defining the key once and then using it multiple times should be faster than the program having to look for it again every time you want to use it.

**Comment by mcollina** [comment]
Regarding var-vs-let, the problem exists in Node v10: https://github.com/fastify/fastify/pull/2667#issuecomment-720592319.

File: lib/contentTypeParser.js [hunk 21]
```diff
@@ -93,19 +93,21 @@ ContentTypeParser.prototype.existingParser = function (contentType) {
...
 
 ContentTypeParser.prototype.getParser = function (contentType) {
-  /* eslint-disable no-var */
-  for (var i = 0; i < this.parserList.length; i++) {
-    if (contentType.indexOf(this.parserList[i]) > -1) {
-      const parser = this.customParsers[this.parserList[i]]
+  // eslint-disable-next-line no-var
+  for (var i = 0, len = this.parserList.length; i !== len; ++i) {
+    const parserName = this.parserList[i]
+    if (contentType.indexOf(parserName) > -1) {
```
Comment (jsumners) [inline lib/contentTypeParser.js:99]:
```suggestion
  for (var i = 0, len = this.parserList.length; i !== len; ++i) {
    const parserName = this.parserList[i]
    if (contentType.indexOf(parserName) > -1) {
```

+ I think the caching of `parserName` is a fine optimization.
+ The pre-calculation of `len` is unnecessary. V8 long ago added this internally.
+ I don't particularly care for either form of the increment operator. I prefer `i += 1`.

Comment (mcollina) [reply lib/contentTypeParser.js:99]: +1

Comment (RafaelGSS) [reply lib/contentTypeParser.js:99]: +1

**Comment by jsumners** [comment]
There's some sort of TS error https://github.com/fastify/fastify/pull/2992/checks?check_run_id=2390180915#step:6:10978

**Review by jsumners — APPROVED** [review]
LGTM

**Comment by Vetlix** [comment]
> There's some sort of TS error https://github.com/fastify/fastify/pull/2992/checks?check_run_id=2390180915#step:6:10978

I’ll check it later, not sure what caused it

**Comment by mcollina** [comment]
@Vetlix please rebase on top of master, it's fixed there.

**Pull request merged**